### ISSUE TITLE
Using 2 GPUs training to train DARTS in parallel, but get 2 different search architecture?

### ISSUE THREAD

**Issue by toufunao** [microsoft/nni#5110]
**Describe the issue**:

I used 2 GPUs to train DARTS, but from the output, I find that I get 2 different results. 

And I used 'export_onnx', but I didn't get the output model under the specified directory.

```python
if __name__ == "__main__":
    parser = ArgumentParser("darts")
    parser.add_argument("--layers", default=8, type=int)
    parser.add_argument("--batch-size", default=64, type=int)
    parser.add_argument("--log-frequency", default=10, type=int)
    parser.add_argument("--epochs", default=1, type=int)
    parser.add_argument("--channels", default=16, type=int)
    parser.add_argument("--unrolled", default=False, action="store_true")
    parser.add_argument("--visualization", default=False, action="store_true")
    parser.add_argument("--v1", default=False, action="store_true")
    parser.add_argument("--local_rank", default=0, type=int,help="node rank for distributed training")
    args = parser.parse_args()
    dataset_train, dataset_valid = datasets.get_dataset("cifar10")
    length=len(dataset_train)
    train_size,validate_size=int(0.5*length),int(0.5*length)
    train_set,validate_set=torch.utils.data.random_split(dataset_train,[train_size,validate_size])
    model = CNN(32, 3, args.channels, 10, args.layers)
    evaluator = pl.Classification(
      train_dataloaders=pl.DataLoader(train_set, batch_size=45,pin_memory=True,num_workers=4),
      val_dataloaders=pl.DataLoader(validate_set, batch_size=45,pin_memory=True,num_workers=4),
      max_epochs=1,
      accelerator="gpu", 
      devices=2,
      strategy='ddp',
      log_every_n_steps=10,
      export_onnx=Path("searched_models/"),
    )
    exploration_strategy = strategy.DARTS()
    exp = RetiariiExperiment(model, evaluator=evaluator, strategy=exploration_strategy)
    exp_config = RetiariiExeConfig('local')
    exp_config.experiment_name = 'cifa10'
    exp_config.max_trial_number = 1  
    exp_config.trial_concurrency = 1 
    exp_config.trial_gpu_number = 2
    exp_config.training_service.use_active_gpu = True

    exp_config.execution_engine = 'oneshot'
    
    exp.run(exp_config, 8081)
    
    exported_arch=exp.export_top_models()[0]
    print(exported_arch)
```

output model:
It seems two processes on 2 GPUs print different results. Should I just use the model trained by rank 0 or is there anything wrong with my use?

```shell
{'normal_n2_p0': 'maxpool', 'normal_n2_p1': 'maxpool', 'normal_n2_switch': [0, 1], 'normal_n3_p0': 'maxpool', 'normal_n3_p1': 'maxpool', 'normal_n3_p2': 'maxpool', 'normal_n3_switch': [0, 1], 'normal_n4_p0': 'maxpool', 'normal_n4_p1': 'maxpool', 'normal_n4_p2': 'maxpool', 'normal_n4_p3': 'maxpool', 'normal_n4_switch': [0, 2], 'normal_n5_p0': 'maxpool', 'normal_n5_p1': 'maxpool', 'normal_n5_p2': 'maxpool', 'normal_n5_p3': 'maxpool', 'normal_n5_p4': 'dilconv5x5', 'normal_n5_switch': [0, 3], 'reduce_n2_p0': 'maxpool', 'reduce_n2_p1': 'maxpool', 'reduce_n2_switch': [0, 1], 'reduce_n3_p0': 'maxpool', 'reduce_n3_p1': 'maxpool', 'reduce_n3_p2': 'maxpool', 'reduce_n3_switch': [0, 2], 'reduce_n4_p0': 'maxpool', 'reduce_n4_p1': 'maxpool', 'reduce_n4_p2': 'maxpool', 'reduce_n4_p3': 'maxpool', 'reduce_n4_switch': [0, 2], 'reduce_n5_p0': 'maxpool', 'reduce_n5_p1': 'sepconv5x5', 'reduce_n5_p2': 'maxpool', 'reduce_n5_p3': 'dilconv5x5', 'reduce_n5_p4': 'dilconv3x3', 'reduce_n5_switch': [0, 3]}
{'normal_n2_p0': 'maxpool', 'normal_n2_p1': 'maxpool', 'normal_n2_switch': [0, 1], 'normal_n3_p0': 'maxpool', 'normal_n3_p1': 'maxpool', 'normal_n3_p2': 'maxpool', 'normal_n3_switch': [0, 1], 'normal_n4_p0': 'maxpool', 'normal_n4_p1': 'maxpool', 'normal_n4_p2': 'sepconv5x5', 'normal_n4_p3': 'dilconv5x5', 'normal_n4_switch': [0, 3], 'normal_n5_p0': 'maxpool', 'normal_n5_p1': 'maxpool', 'normal_n5_p2': 'maxpool', 'normal_n5_p3': 'maxpool', 'normal_n5_p4': 'maxpool', 'normal_n5_switch': [0, 3], 'reduce_n2_p0': 'maxpool', 'reduce_n2_p1': 'maxpool', 'reduce_n2_switch': [0, 1], 'reduce_n3_p0': 'maxpool', 'reduce_n3_p1': 'maxpool', 'reduce_n3_p2': 'maxpool', 'reduce_n3_switch': [0, 2], 'reduce_n4_p0': 'maxpool', 'reduce_n4_p1': 'maxpool', 'reduce_n4_p2': 'maxpool', 'reduce_n4_p3': 'maxpool', 'reduce_n4_switch': [0, 2], 'reduce_n5_p0': 'maxpool', 'reduce_n5_p1': 'dilconv5x5', 'reduce_n5_p2': 'dilconv5x5', 'reduce_n5_p3': 'dilconv5x5', 'reduce_n5_p4': 'sepconv5x5', 'reduce_n5_switch': [0, 3]}
```

![output_model](https://user-images.githubusercontent.com/39375136/188383094-55694fff-c595-48fa-a716-78de95e24a0d.jpeg)

**Comment by ultmaster** [comment]
I just realized that the architecture parameters might not be synchronized across devices because of the hacky `parameters()`. This needs to be checked.

However, this might be caused by another reason: batch norm is not properly synchronized across devices.

**Comment by Yuanyang-Zhu** [comment]
If only changes are made on device 0, other devices will execute the original parameters. If the device is not restricted, NNI will be stuck waiting for other device parameters
